wireup/ioc/validation.py

```diff
@@ -94,6 +95,37 @@ def assert_dependency_exists(container: BaseContainer, parameter: AnnotatedParam
         raise WireupError(msg)
 
 
+def assert_valid_resolution_path(
+    container: BaseContainer, dependency: AnnotatedParameter, path: list[tuple[AnnotatedParameter, Any]]
+) -> None:
+    """Assert that the resolution path for a dependency does not create a cycle."""
+    if dependency.klass in container._registry.interfaces or dependency.is_parameter:
+        return
+    dependency_service_factory = container._registry.factories.get((dependency.klass, dependency.qualifier_value))
+    if any(p.klass == dependency.klass and p.qualifier_value == dependency.qualifier_value for p, _ in path):
+
+        def stringify_dependency(p: AnnotatedParameter, factory: Any) -> str:
```

Comment:
Factory and qualifier aren't mutually exclusive so we'd need to mention both if they're present.